src/Compiler/Facilities/AsyncMemoize.fs

```diff
@@ -153,409 +141,108 @@ type internal AsyncMemoize<'TKey, 'TVersion, 'TValue when 'TKey: equality and 'T
-                    internalError key.Label $"Invalid state: Failed Completed job \n%A{ex}"
-
-                | JobCompleted(_result, _diags), Some(Completed(_job, _diags2)) ->
-                    internalError key.Label "Invalid state: Double-Completed job"
-
-                | JobFailed(ex, _diags), Some(Job.Failed(_, ex2)) ->
-                    internalError key.Label $"Invalid state: Double-Failed job \n%A{ex} \n%A{ex2}"
-
-                | JobCompleted(_result, _diags), Some(Job.Failed(_, ex2)) ->
-                    internalError key.Label $"Invalid state: Completed Failed job \n%A{ex2}"
-            })
-
-    let rec post msg =
-        Interlocked.Increment &updates_in_flight |> ignore
-        backgroundTask {
-            do! processStateUpdate post msg
-            Interlocked.Decrement &updates_in_flight |> ignore
-        }
-        |> ignore
-
-    member this.Get'(key, computation) =
-
-        let wrappedKey =
-            { new ICacheKey<_, _> with
-                member _.GetKey() = key
-                member _.GetVersion() = Unchecked.defaultof<_>
-                member _.GetLabel() = match key.ToString() with | null -> "" | s -> s
-            }
-
-        this.Get(wrappedKey, computation)
+                | CacheEvent.Evicted -> logK JobEvent.Evicted
+                | CacheEvent.Collected -> logK JobEvent.Collected
+                | CacheEvent.Weakened -> logK JobEvent.Weakened
+                | CacheEvent.Strengthened -> logK JobEvent.Strengthened
+                | CacheEvent.Cleared -> logK JobEvent.Cleared))
 
     member _.Get(key: ICacheKey<_, _>, computation) =
-
         let key =
             {
                 Label = key.GetLabel()
                 Key = key.GetKey()
                 Version = key.GetVersion()
             }
 
-        async {
-            let! ct = Async.CancellationToken
+        let wrappedComputation =
+            async {
+                use! _handler = Async.OnCancel (fun () -> log Canceled key)
+                let sw = Stopwatch.StartNew()
+                log Started key
+                let logger = CapturingDiagnosticsLogger "cache"
+                SetThreadDiagnosticsLoggerNoUnwind logger
 
-            let callerDiagnosticLogger = DiagnosticsThreadStatics.DiagnosticsLogger
+                try
+                    let! result = computation
+                    log Finished key
+                    Interlocked.Add(&duration, sw.ElapsedMilliseconds) |> ignore
+                    return Result.Ok result, logger
+                with
+                | ex ->
+                    log Failed key
+                    return Result.Error ex, logger
+            }
 
-            match!
-                processRequest post (key, GetOrCompute(computation, ct)) callerDiagnosticLogger
-                |> Async.AwaitTask
-            with
-            | New internalCt ->
+        let getOrAdd () =
+            let cached, otherVersions = cache.GetAll(key.Key, key.Version)
```

Comment:
So now when a job errors out, it gets cached and if re-requested we just give back the error? I think it might be better to restart in that case. There's always a chance I/O is involved and the error was caused by some transient issue. Maybe this could be configurable.

Reply:
What I was not sure of here: 
Are any of the cached jobs in TransparentCompiler _expected_ to throw exceptions in some situations or is it always handled by the `ErrorR`, diagnostics logging and so on? If the exception here means internal error, this should not be cached.

We could also easily make this configurable per cache or even per exception type. For example remove the job from the cache immediately or after some reasonably short timespan. Also possible to add a `resetOnFailure` flag to AsyncLazy.

Reply:
Good point. I think at the moment not and we should probably make it a design decision to keep it that way. If it's an expected "failure" it should be encoded in the result not thrown as exception. Still makes sense to make it configurable per cache and on by default.